SharpRepository.EfRepository/EfRepositoryBase.cs

```diff
@@ -33,7 +34,7 @@ protected override void AddItem(T entity)
             if (typeof(TKey) == typeof(Guid) || typeof(TKey) == typeof(string))
             {
                 TKey id;
-                if (GetPrimaryKey(entity, out id) && Equals(id, default(TKey)))
+                if (GetPrimaryKey(entity, out id) && GenerateKeyOnAdd && Equals(id, default(TKey)))
```

Comment:
Let's move the check for GenerateKeyOnAdd to the first one since it would be the quickest to fail